sdk/python/feast/infra/aws.py

```diff
@@ -306,21 +307,52 @@ def _create_or_get_repository_uri(self, ecr_client):
 
 def _get_lambda_name(project: str):
     lambda_prefix = AWS_LAMBDA_FEATURE_SERVER_REPOSITORY
-    lambda_suffix = f"{project}-{_get_version_for_aws()}"
+    lambda_suffix = f"{project}-{_get_docker_image_version()}"
     # AWS Lambda name can't have the length greater than 64 bytes.
-    # This usually occurs during integration tests or when feast is
-    # installed in editable mode (pip install -e), where feast version is long
+    # This usually occurs during integration tests where feast version is long
     if len(lambda_prefix) + len(lambda_suffix) >= 63:
         lambda_suffix = base64.b64encode(lambda_suffix.encode()).decode()[:40]
     return f"{lambda_prefix}-{lambda_suffix}"
 
 
-def _get_version_for_aws():
-    """Returns Feast version with certain characters replaced.
+def _get_docker_image_version() -> str:
+    """Returns a version for the feature server Docker image.
+
+    For public Feast releases this equals to the Feast SDK version modified by replacing "." with "_".
+    For example, Feast SDK version "0.14.1" would correspond to Docker image version "0_14_1".
+
+    For integration tests this equals to the git commit hash of HEAD. This is necessary,
+    because integration tests need to use images built from the same commit hash.
+
+    During development (when Feast is installed in editable mode) this equals to the Feast SDK version
+    modified by removing the "dev..." suffix and replacing "." with "_". For example, Feast SDK version
+    "0.14.1.dev41+g1cbfa225.d20211103" would correspond to Docker image version "0_14_1". This way,
+    Feast SDK will use an already existing Docker image built during the previous public release.
 
-    This allows the version to be included in names for AWS resources.
     """
-    return get_version().replace(".", "_").replace("+", "_")
+    if flags_helper.is_test():
```

Comment:
Might be worth having a comment that that calls out that changes to this version for test should also impact the workflow yaml